lib/engine/game/g_18_uruguay/game.rb

```diff
@@ -0,0 +1,908 @@
+
+        def route_include_port?(route)
+          route.hexes.each do |hex|
+            return true if PORTS.include?(hex.id)
+          end
+          false
+        end
+
+        def reset_train!(_train)
+          p 'Reset train!'
+          # train.name = train.name.partition('+')[0] unless train.nil?
+        end
+
+        def check_for_goods_if_run_to_port(route, visits)
+          true if route.corporation == @rptla
+          visits_include_port?(visits) || !train_with_goods?(route.train)
+        end
+
+        def check_for_port_if_goods_attached(route, visits)
+          true if route.corporation == @rptla
+          !visits_include_port?(visits) || train_with_goods?(route.train)
+        end
+
+        def check_distance(route, visits, train = nil)
+          @round.current_routes |= [route]
+          if route.corporation != @rptla && !nationalized?
+            raise RouteTooLong, 'Need to have goods to run to port' unless check_for_port_if_goods_attached(route,
+                                                                                                            visits)
+            raise RouteTooLong, 'Goods needs to be shipped to port' unless check_for_goods_if_run_to_port(route,
+                                                                                                          visits)
+          end
+          raise RouteTooLong, '4D trains cannot deliver goods' if route.train.name == '4D' && visits_include_port?(visits)
+
+          super
+        end
+
+        def operating_order
+          super.sort.partition { |c| c.type != :bank }.flatten
+        end
+
+        def float_corporation(corporation)
+          return if corporation == @rptla
+          return unless @loans
+
+          amount = corporation.par_price.price * 5
+          @bank.spend(amount, corporation)
+          @log << "#{corporation.name} receives #{format_currency(corporation.cash)}"
+          take_loan(corporation, @loans[0]) if @loans.size.positive? && !nationalized?
+        end
+
+        def second_capitilization!(corporation)
+          amount = corporation.par_price.price * 5
+          @bank.spend(amount, corporation)
+          @log << "Connected to destination #{corporation.name} receives #{format_currency(amount)}"
+        end
+
+        def after_buy_company(player, company, _price)
+          abilities(company, :shares) do |ability|
+            ability&.shares&.each do |share|
+              return super unless share.corporation == @rptla
+
+              share_pool.buy_shares(player, share, exchange: :free)
+              after_par(share.corporation) if share.president
+              ability.use!
+            end
+          end
+        end
+
+        def after_par(corporation)
+          super
+          return unless corporation == @rptla
+
+          train = find_and_remove_train_by_id('Ship 1-0')
+          buy_train(@rptla, train)
+        end
+
+        # LOANS
+        def init_loans
+          @loan_value = 100
```

Comment:
Could be pulled out to the top of the file as a constant instead of being an instance variable that is re-initialized every time this function is called?